app/src/main/java/ch/hikemate/app/model/elevation/ElevationServiceRepository.kt

```diff
@@ -1,113 +1,265 @@
+
+/**
+ * A data class that represents a cache entry for the elevation data
+ *
+ * @param timestamp The timestamp of the cache entry (i.e when the data was added to the cache)
+ * @param elevation The elevation data
+ */
+data class ElevationCacheEntry(val timestamp: Date, val elevation: Double)
+
+/**
+ * A request for the ElevationService
+ *
+ * @param coordinates The list of coordinates to get the elevation of
+ * @param onSuccess The callback to be called when the request is successful
+ * @param onFailure The callback to be called when the request fails
+ */
+data class ElevationRequest(
+    val coordinates: List<LatLong>,
+    val onSuccess: (List<Double>) -> Unit,
+    val onFailure: (Exception) -> Unit
+)
 
 /**
  * A repository for the ElevationService. This class is responsible for making the network request
  */
-class ElevationServiceRepository(private val client: OkHttpClient) : ElevationService {
+class ElevationServiceRepository(
+    private val client: OkHttpClient,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
+) : ElevationRepository {
+
+  // The number of failed requests made to the Elevation API since the last successful request
+  private var failedRequests = 0
 
   // Cache for the elevation data, indexed by hike ID
-  private val cache = mutableMapOf<String, List<Double>>()
+  private val cache = hashMapOf<LatLong, ElevationCacheEntry>()
+
+  // List of requests that are waiting for the cache to be updated
+  private val requests = mutableListOf<ElevationRequest>()
+
+  // A mutex to ensure that the cache and requests are accessed atomically
+  private val mutex: Mutex = Mutex()
+
+  // The job performing the cache update
+  private var cacheUpdateJob: Job? = null
 
   /**
    * Get the size of the cache
    *
    * @return The size of the cache
    */
-  fun getCacheSize(): Int {
+  private fun getCacheSize(): Int {
     return cache.size
   }
 
   override fun getElevation(
       coordinates: List<LatLong>,
-      hikeID: String,
       onSuccess: (List<Double>) -> Unit,
       onFailure: (Exception) -> Unit
   ) {
     if (coordinates.isEmpty()) {
       onSuccess(emptyList())
       return
     }
-    // If the cache already has the elevation data for this hike, call onSuccess
-    if (cache.containsKey(hikeID)) {
-      cache[hikeID]?.let { onSuccess(it) }
+
+    if (coordinates.minus(cache.keys).isEmpty()) {
+      // All coordinates are in the cache
+      onSuccess(
+          coordinates.map {
+            val value = cache[it]
+            if (value == null) {
+              Log.e("ElevationServiceRepository", "Coordinate is missing in cache")
+              return@map 0.0
+            } else {
+              value.elevation
+            }
+          })
       return
     }
 
-    // Check if a request is already in progress
-    if (cache[hikeID] == null) {
-      cache[hikeID] = emptyList()
-    } else {
-      return
+    CoroutineScope(dispatcher).launch {
```

Comment:
It totally could! Do you want me to keep the class level one if we don't provide one or do you want me to remove it completely?

Reply:
In the end, did you change anything more than the name of the class-level dispatcher parameter?

I don't think doing more is necessary though. If we pass a scope to each function, it would probably be the view model's scope. However, the view model is created in `MainActivity`, so using its scope stays valid even when changing between screens.

This means that using its scope would not stop requests unless the app is left by the user, but in that case the requests will be stopped anyway, so no need for this.